F#182 partial trajectory mod

beatrizleon opened the pull request:
Enables the control of partial trajectories. In order to do this, a trajectory now is composed of a vector of vectors of segments (one vector of segments per each joint), so each joint can a vector with different number of segments. 

The changes of code are the minimum I could modify to make it work. If the changes are accepted, we can improve it. Specially, the definition of a segment, as it is only per joint, it does not need a permutation vector, and all its related methods.

All test are passing, but I had to modify some of them to the new implementation. If it is accepted, we can think of new test needed specifically to test partial trajectories.

Commit:
moving but only with one goal

Commit:
working with one goal, still needs to set the goal handlers correctly

Commit:
Trial with vector of goals but it turns out that the action server can only hold one goal at a time

Commit:
Partial trajectory working. Needs cleaning and check tolerances

Commit:
Working and cleaned

Commit:
Working and cleaned

Commit:
Changes made after modifing tests

Commit:
working on tests

Commit:
Fixed test and all passed

Commit:
Cleaning

adolfo-rt commented:
I did a first pass reading the PR. I'll do a second, more detailed one soon and come back with any comments that might arise. I'd also like to do some experimental testing.

Thanks for putting this together.

Commit:
Fixed the import of new trajectories with accelerations empty and a test

beatrizleon commented:
Thanks @adolfo-rt for reviewing it. I found an error in my branch when a new trajectory was send with an empty vector of accelerations. It is fixed now.

Commit:
Small fix to error message

adolfo-rt commented on joint_trajectory_controller/include/joint_trajectory_controller/init_joint_trajectory.h:
```diff
@@ -114,6 +113,12 @@ struct InitJointTrajectoryOptions
+template <class Trajectory>
+bool IsNotEmpty(typename Trajectory::value_type trajPerJoint)
+{
+  return trajPerJoint.size()>0;
```
`return !trajPerJoint.empty();`

adolfo-rt commented:
@beatrizleon Update: I'm in the middle of a detailed code review. I'm about halfway done, and I'd like to give some feedback tomorrow.

adolfo-rt commented on joint_trajectory_controller/include/joint_trajectory_controller/init_joint_trajectory.h:
```diff
@@ -114,6 +113,12 @@ struct InitJointTrajectoryOptions
 };
 
+template <class Trajectory>
+bool IsNotEmpty(typename Trajectory::value_type trajPerJoint)
```
`isNotEmpty`

adolfo-rt commented on joint_trajectory_controller/include/joint_trajectory_controller/init_joint_trajectory.h:
```diff
@@ -295,93 +313,156 @@ Trajectory initJointTrajectory(const trajectory_msgs::JointTrajectory&       msg
-        ROS_ERROR("Cannot create trajectory from message. "
-                  "Vector specifying whether joints wrap around has an invalid size.");
-        return Trajectory();
+        (result_traj[joint_id])[segment_id].setGoalHandle(options.rt_goal_handle);
```
The new goal handle should not be applied to the trajectory currently being executed. Only to the segments of the new one, and to the bridge segment that continuously transitions between the current and new trajectories, which is done below.

beatrizleon replied:
I thought about implementing it that way at first and have a saved list of active goals, however, I realized that the action server can handle only one goal at a time (http://wiki.ros.org/actionlib). As thechnically the other joints need to finish their trajectories, what I did was prempted the previous goal and creating a new goal (as it was originally), but now the new goal contains the trajectories for all joints, only that each joint has different number of segments according to the partial trajectories. Then the goal will be considered finished when all joints finish their corresponding trajectories. I have tested this behaviour with our hand and it is working as expected.

adolfo-rt replied:
Still, why are you assigning the new goal handle to the previous trajectory segments?.

One thing that would simplify the review process is some sort of design explanation of what technical problems you've faced, and the proposed solutions. This could go in the PR description. This helps a lot to review large PRs.

A complementary strategy is to have a commit history that tackles each problem individually, so each commit can be reviewed almost independently. Sometimes this is not possible, because of the way individual technical problems are coupled.

beatrizleon replied:
Agree. I fixed it to only assign it to segments after current time.

adolfo-rt commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:
```diff
@@ -393,48 +347,100 @@ update(const ros::Time& time, const ros::Duration& period)
-    return;
-  }
+  std::vector<bool> successful_joint_traj;
+  successful_joint_traj.resize(joints_.size(), false);
```
Dynamic memory allocations are disallowed in the real-time path.

In controllers, computation takes place in two places: 
- `update` method: contains periodic, real-time computation.
- ROS API callbacks: contains asynchronous, non real-time computation.

When implementing the `update` method, one has to be very careful not to break determinism and real-time friendliness.

adolfo-rt replied:
You can likely make the method real-time safe by pre-allocating resources on `init`, as the sizes are known at initialization time.

I forgot to say above that `starting` and `stopping` are also in the real-time path.

adolfo-rt commented on joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:
```diff
@@ -393,48 +347,100 @@ update(const ros::Time& time, const ros::Duration& period)
   // Update current state and state error
   for (unsigned int i = 0; i < joints_.size(); ++i)
   {
+    typename Segment::State desired_joint_state_ = typename Segment::State(1);
```
Dynamic allocations are taking place here and in the line below as well.

adolfo-rt commented on joint_trajectory_controller/include/joint_trajectory_controller/init_joint_trajectory.h:
```diff
@@ -66,9 +66,6 @@ inline std::vector<unsigned int> permutation(const T& t1, const T& t2)
   typedef unsigned int SizeType;
 
-  // Arguments must have the same size
-  if (t1.size() != t2.size()) {return std::vector<SizeType>();}
```
If vectors are no longer to be of the same size, then we are no longer talking about a permutation (rearranging). What's being computed is a map between `t1` indices (implicitly encoded in return vector indices) to `t2` indices. Also, the size of `t1` cannot be larger than that of `t2`. That would be an early optimization.

The function name and documentation should be changed accordingly.

beatrizleon replied:
I agree. What name should be better? maybe "mapping" or "jointMapping"?

beatrizleon replied:
Fixed it and added a test.

beatrizleon replied:
I have also removed the permutation variable inside a state and segment as they are now a set for a specific joint.

adolfo-rt commented:
I ran the tests and tested it in simulation. Seems to work as expected (yay).

**Make it opt-in**
One important thought:  Partial joint specification should be opt-in and disabled by default. Rationale:

When a new command specifying a subset of the joints is processed, the unspecified joints keep on executing their previous goal as if nothing happened. I see potential for robots doing unexpected things when different commands that don't know about each other interact. There's always been the risk of controller clients sending bogus commands and crashing the robot or some other unexpected behavior, but the interaction of conflicting commands within a controller was never an issue. I would consider this an advanced behavior that has to be enabled explicitly, as in _I know what I'm doing_.

**Review thoroughness**
Review-wise, I have to confess that the changes are so extensive, that even after having spent a few hours reviewing them, I don't get the feeling of having been thorough enough. It's good that tests are passing, although I don't feel very confident about the completeness of the tolerance checking tests (my fault). When I wrote the tests, that was the only part that was not exercised to the fullest.

To make the test suite more complete and exercise the new feature, I'm thinking of: while the controller executes a trajectory involving all joints (not holding position, but actually moving), send a new command (topic interface) or goal (action interface) that:
- Send trajectories with all waypoints in the past: Nothing gets executed.
- Send trajectories with some waypoints on the past, and some in the future: Some get executed.
- Send trajectories with all waypoints in the future. All get executed, in the future.

I would also like the implementation to be validated on a real-time system.

**More**

The structure and style of the patch can be improved for better code reading and brevity ,but I'll skip those comments for now.

beatrizleon commented:
@adolfo-rt  Thanks for the review. I am going through your comments now.

Commit:
Fixed comments

Commit:
Removed permutation from vector class and fixed tests

Commit:
Fixed comment about assigning goal handle only to segments after current time

beatrizleon commented:
How are you thinking we could implement the opt-in option?

adolfo-rt commented:
> How are you thinking we could implement the opt-in option? 

With a ROS parameter, say `allow_partial_joint_specification`, or something shorter, but equally descriptive. One would have to explicitly set it to `True` to enable partial joint specification. To enforce the behavior, you would have to do something similar to what was done up to now:
- When using the action inteface. Keep the [check](https://github.com/ros-controls/ros_controllers/blob/indigo-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L545-L552) that rejected goals early on, but disable it if `allow_partial_joint_specification == True`
- For topic commands, the check that kicked-in was [this one](https://github.com/ros-controls/ros_controllers/blob/indigo-devel/joint_trajectory_controller/include/joint_trajectory_controller/init_joint_trajectory.h#L248-L252). Same story, keep the check, but disable it if `allow_partial_joint_specification == True`.

Note that these checks relied on the `permutation` function, which actually checked for permutations (not mappings). As an alternative, you can call your new mapping function and additionally check that the two input vectors have the same size.

Commit:
Fixed setting of success to goal handle when all joints succeeded

Commit:
Added rosparam to allow partial joints goals

beatrizleon commented:
Added the ros param. I named it: "allow_partial_joints_goal" but it is opened to suggestions :)

Commit:
added tests for partial joint goals

beatrizleon commented:
@adolfo-rt I added a set of tests (very similar to the existing ones) to test partial trajectories which included the ones you mentioned in your previous comment.

We do not have a real time system, then I would not be able to test that.

Appart from that I think I have addressed all your comments. Let me know what you think.

ugocupcic commented:
@adolfo-rt any news on this? We're kind of waiting for that functionality at Shadow :)

Cheers,

Ugo

sachinchitta commented:
@ugocupcic - @adolfo-rt did mention on the mailing list that he would be offline for 3 weeks starting Sept. 8.

ugocupcic commented:
Ah! OK thanks @sachinchitta :)

toliver commented:
@adolfo-rt any new on this? If there is anything else that we could improve to get it merged please let us know.

v4hn commented:
It has been nearly three month since @adolfo-rt was last active on github.
Is there anyone else who could look at that?
It was mentioned that this should be in the kinetic release in https://github.com/ros-controls/ros_control/issues/234 . Is there any reason why this should not be merged into indigo as well?
It would help me quite a lot if it were.

bmagyar commented:
I made the releases for Kinetic and if all goes fine (since it's a well discussed issue) I will merge it to kinetic-devel soon. I'm not sure though if these changes should go to Indigo&Jade as well, so my approach so far was to merge all new stuff to kinetic-devel, and wait for some of our sleeping agents to appear so we can discuss propagating them back to indigo&jade.

toliver commented:
We are going to try to rebase this branch and open a new PR against kinetic-devel. Hopefully this week or the next.

@bmagyar  please let us know if we should also keep this PR open just in case it's getting merged into indigo later on.

bmagyar commented:
Great Toni!

For the rebase may I ask you to also either squash the commits into one huge commit or into bigger chunks so that the "work in progress" type of commits are not reflected in the proposed changes? It's better if we track feature additions rather than the entire development processes.

Once the new PR is open, this can be closed. I am personally ok to push through all the new stuff for ROS Kinetic but if you'd like back-support for Indigo we will have to gather some more people and get a lot more feedback. If you wish to have the Kinetic features in older distros, feel free to open an issue and let's discuss them there.

bmagyar commented:
Closing since we merged it to Kinetic.
